### ISSUE TITLE
PDF Reading in-app, not download and open local (android/ios)

### ISSUE THREAD

**Issue by BloodyIron** [nextcloud/android#4802]
I really want to use the nextCloud app on Android to read PDFs and have that app do the PDF rendering. Namely so that my progress in PDF files (and other docs, like epub, or whatever) are kept server-side, instead of on my Android device.

Considering the web interface already does PDF rendering instead of downloading and opening local, I think this would be an extension of an already reasonable expectation, that PDF rendering is done in the client (web, and in this case android/ios).

Any chance we can get this to be a thing? Please? :)

I'm sure iOS users would appreciate the same thing too (I don't have any iOS devices).

**Comment by tobiasKaminsky** [comment]
Hm, I do not fully understand the benefit of having a PDF view internal in the app.
You always have to download the PDF entirely (as right now).

**Comment by BloodyIron** [comment]
I don't know how to better explain my angle here. I switch between devices regularly for reading content and I want my position in that document to be server-side. Instead of having to re-remember where I left off from one device to the next. :/

**Comment by tobiasKaminsky** [comment]
Ah, now I get what you want.
Unfortunately this is not something we have, or which the server supports.
Also it is in my opinion out of scope.

@AndyScherzinger

**Comment by BloodyIron** [comment]
How is it out of scope? This is a feature request for a new feature, of course it's not something it can do today...

**Comment by maxolasersquad** [comment]
The primary purpose of the NC app is file sync, not a thick-client replacement for the web interface. Which is why this feature is beyond the scope of the app 

What you want might be achievable on the web interface with the ebook reader extension. 

https://apps.nextcloud.com/apps/files_ebookreader

**Comment by BloodyIron** [comment]
The web interface already can do it, but I don't see how PDF reading is an unacceptable feature to add. It's not going to make the application bloated and unwieldy, it will improve the UX in multiple regards, primarily that your progress in documents is consistently tracked from one User Interface to the next (phone to desktop).

nextCloud as a virtual library is an awesome tool, and not having position tracking that is consistent breaks that.

**Comment by loegstrup** [comment]
I agree that having an in-app PDF reader could (1) enhance the user experience as BloodyIron mentions and (2) fix security concerns caused by the PDF-files having to be stored on the local device.

To illustrate how the app could currently leave a false sense of security:
- Go to settings
- Choose an app password
- Close NextCloud
- Only way you **believe** that PDF-files can be accessed is by opening NextCloud and typing in your password
- This is wrong, since all opened PDF-files can be accessed by going to 'Files' and (1) accessing them in the NextCloud directory or (2) in some cases, simply looking in the latest accessed files

Of course being in a situation where someone else can access your 'Files' seems like bad practice to begin with, but that is not the point. The point is that by relying solely on Android to handle file preview, you also rely on the operating system to do a proper job of respecting the apps privacy settings, which could potentially lead to someone breaching data without even knowing it (it is not everyone who understands the green checkmark indicating that the file has now been put in the local storage).

**Comment by tobiasKaminsky** [comment]
If we implement PDF viewer, then still the file will be needed to download (and stored)
If you want to not expose your synced files, then please change storage location to /data/user/0/com.nextcloud.client/files and then only this app can access the files

**Comment by BloodyIron** [comment]
While clearly the app needs some form of cached or local copy of the PDF, would not it be possible to have the app keep this in a secured enclave unless "allowed" for the user to download locally? This could be used as a way for data exfiltration otherwise, and that is a problem in IT.

**Comment by msviba** [comment]
I would like to migrate from Dropbox. There is feature of internal pdf viewer.
For example:
Received email has attached pdf file. I will open file and chose Nextcloud app as viewer. Pdf will be open in nextcloud app with context menu where is ability to chose folder and upload to NC 

This is one click solution how to sync pdf from email. Dropbox has it and it cool.

**Comment by BloodyIron** [comment]
Yeah this behaviour (download and not read in the android nextCloud app) still exists. Do these feature requests EVER get read? It's been almost 4 years now! >:|

**Comment by msviba** [comment]
> Yeah this behaviour (download and not read in the android nextCloud app) still exists. Do these feature requests EVER get read? It's been almost 4 years now! >:|

No, it doesn't exist. Please try it with some attached pdf file in gmail app. I have android 10 (Huawei) and it's not possible to save it into Nextcloud.

**Comment by BloodyIron** [comment]
Ahem, I meant to say... still _does not_ exist XD that's an accidental typo on my end, whoops!

**Comment by Deft-54** [comment]
I just want to say I love the idea of reading my pdf without downloading them... And I keep searching every month if I would be add. I Hope it would soon 🤞

**Comment by GlassedSilver** [comment]
Guys... All I'm gonna say is that if this is out of scope, but the project competes with the likes of Google Drive, then there is a bit of an underappreciation for excellence in good and complex (behind-the-scenes) UX and UI.

The client to this day feels very much like your typical sync client, much less like a window to your personal cloud. Between auto-uploads being janky and unable to delete local files and things like "Send usage reports" requests not being able to be turned off server-wide, there are things like these that make me question if the goal is ACTUALLY to give big tech a run for its money.

Google doesn't have office or pdf viewing in the Drive app, but it will deeplink you into the Docs app for example that loads the document right away.

Right now the workflow is to open the document, have Collabora for Android installed which is a rather so-and-so app... and then you save your changes to your locally stored version which then... hopefully syncs back to the server.

Unless I'm missing something here of course.

IMHO that's a rather archaic way to handle it. Ideally Nextcloud would hand off the URL to the office app of choice and its edits go straight to the server much like how the webapp handles things. Of course working on the locally cached or downloaded file should ALWAYS remain a possibility, I'm only talking about the default behavior.

As for PDF specifically, which this issue is actually is about, IDK which apps can integrate with Nextcloud tightly, but another issue is that they all need to oAuth separately to the server rather than being able to share the authorization as Google's apps do. I think this mostly stems from the fact that as long as apps are distributed by one entity they can all share private app data which includes sessions, cookies and such. Same thing why Amazon logs me into Prime Video automatically if the Amazon shopping app is already installed and logged-in on that device.
